### PR TITLE
Big float limitations

### PR DESCRIPTION
**Pull request by andrei-diaconescu**

### CODE REVIEW

File: test/contracts/big-floats-2/big-floats-2.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
+#include "../elrond/context.h"
+#include "../elrond/bigInt.h"
+#include "../elrond/types.h"
+#include "../elrond/test_utils.h"
+#include "../elrond/bigFloat.h"
+
+// byte gobEncodedFloat1[] = {1, 10, 0, 0, 0, 100, 0, 0, 0, 108, 136, 217, 65, 19, 144, 71, 160, 0};
+// // = 173476272346174583562347456134583.6134671346713451345
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply test/contracts/big-floats-2/big-floats-2.c:8]: done

File: test/contracts/big-floats-2/big-floats-2.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+void BigFloatNewFromFracTest()
+{
+    int numerator = int64getArgument(0);
+    int denominator = int64getArgument(1);
+    int bigFloatHandle = bigFloatNewFromFrac(numerator, denominator);
+}
+
+void BigFloatNewFromSciTest()
+{
+    int significand = int64getArgument(0);
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply test/contracts/big-floats-2/big-floats-2.c:35]: https://en.wikipedia.org/wiki/Floating-point_arithmetic

not a romanian name

File: arwen/elrondapi/manBufOps.go [hunk 3]
```diff
@@ -508,15 +510,23 @@ func v1_4_mBufferToBigFloat(context unsafe.Pointer, mBufferHandle, bigFloatHandl
...
-	metering.UseGas(gasToUse)
+	metering.UseAndTraceGas(gasToUse)
 
 	managedBuffer, err := managedType.GetBytes(mBufferHandle)
 	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
 		return 1
 	}
 
+	managedType.ConsumeGasForBytes(managedBuffer)
+
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply arwen/elrondapi/manBufOps.go:524]: done

File: arwen/contexts/managedType.go [hunk 4]
```diff
@@ -259,8 +265,20 @@ func (context *managedTypesContext) newBigIntNoCopy(value *big.Int) int32 {
 
 // BIG FLOAT
 
-func (context *managedTypesContext) GetBigFloatPrecision() uint {
-	return bigFloatPrecision
+func (context *managedTypesContext) BigFloatPrecIsNotValid(precision uint) bool {
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply arwen/contexts/managedType.go:268]: done

File: arwen/contexts/managedType.go [hunk 5]
```diff
@@ -259,8 +265,20 @@ func (context *managedTypesContext) newBigIntNoCopy(value *big.Int) int32 {
...
+	return precision != bigFloatPrecision
+}
+
+func (context *managedTypesContext) BigFloatExpIsNotValid(exponent int) bool {
+	return exponent < bigFloatMinExponent || exponent > bigFloatMaxExponent
+}
+
+func (context *managedTypesContext) EncodedBigFloatIsNotValid(encodedBigFloat []byte) bool {
+	return !bytes.Equal(encodedBigFloat[:6], positiveEncodedBigFloatPrefix[:]) && !bytes.Equal(encodedBigFloat[:6], negativeEncodedBigFloatPrefix[:])
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply arwen/contexts/managedType.go:277]: As far as I've studied and researched, an encoded big float cannot have a length smaller than 6. But for extra safety, I added this verification, also fixing a bug for the 0 value.

File: arwen/elrondapi/bigFloatOps.go [hunk 6]
```diff
@@ -519,7 +556,7 @@ func v1_4_bigFloatPow(context unsafe.Pointer, destinationHandle, opHandle, expon
 
 	//this calculates the length of the result in bytes
 	lengthOfResult := big.NewInt(0).Div(big.NewInt(0).Mul(op2BigInt, big.NewInt(int64(opBigInt.BitLen()))), big.NewInt(8))
-	managedType.ConsumeGasForThisBigIntNumberOfBytes(lengthOfResult)
+	managedType.ConsumeGasForThisBigIntNumberOfBytes(new(big.Int).Abs(lengthOfResult))
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply arwen/elrondapi/bigFloatOps.go:559]: done

File: arwen/contexts/managedType.go [hunk 7]
```diff
@@ -317,6 +352,10 @@ func (context *managedTypesContext) PutBigFloat(value *big.Float) int32 {
 	if value == nil {
 		value = big.NewFloat(0)
 	}
+	exponent := value.MantExp(nil)
+	if exponent > 65025 || exponent < -65025 {
+		value = big.NewFloat(0)
```
[the remark this answers is not in the document]
Comment (andrei-diaconescu) [reply arwen/contexts/managedType.go]: done

**Review by sasurobert — APPROVED** [review]

File: test/features/basic-features/mandos/big_floats.scen.json [hunk 8]
```diff
@@ -1,5 +1,6 @@
 {
     "gasSchedule": "v3",
+    "traceGas": true,
```
Comment (andrei-marinica) [inline test/features/basic-features/mandos/big_floats.scen.json:3]: This is for debugging only, should be turned off when publishing test.

File: arwen/contexts/managedType.go [hunk 9]
```diff
@@ -259,8 +266,26 @@ func (context *managedTypesContext) newBigIntNoCopy(value *big.Int) int32 {
...
+
+// BigFloatExpIsNotValid checks if the exponent of a big float is not valid (smaller than -65025 or bigger than 65025)
+func (context *managedTypesContext) BigFloatExpIsNotValid(exponent int) bool {
+	return exponent < bigFloatMinExponent || exponent > bigFloatMaxExponent
+}
+
+// EncodedBigFloatIsNotValid checks if an encoded big float is not valid
+func (context *managedTypesContext) EncodedBigFloatIsNotValid(encodedBigFloat []byte) bool {
+	length := len(encodedBigFloat)
+	if length < 6 {
```
Comment (andrei-marinica) [inline arwen/contexts/managedType.go:282]: Magic number 6 should be extracted to constant

**Review by camilbancioiu — APPROVED** [review]

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: test/features/basic-features/mandos/big_floats.scen.json (hunk 8 above)
Comment (sasurobert) [reply test/features/basic-features/mandos/big_floats.scen.json:3]: done

**Review by andrei-marinica — DISMISSED** [review]

**Review by sasurobert — DISMISSED** [review]

File: arwen/contexts/managedType.go [hunk 10]
```diff
@@ -37,6 +37,8 @@ const p256CurveUnmarshalCompressedMultiplier = 100
 const p384CurveUnmarshalCompressedMultiplier = 200
 const p521CurveUnmarshalCompressedMultiplier = 400
 
+const minEncodedBigFloatNumArgs = 6
```
Comment (camilbancioiu) [inline arwen/contexts/managedType.go:40]: It's rather the minimum length of a single encoded big float, in bytes. Consider the following rename:
```suggestion
const minEncodedBigFloatLength = 6
```

**Review by camilbancioiu — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**